tox.ini

```diff
@@ -27,6 +23,13 @@ deps =
     url-matcher==0.2.0
     web-poet==0.7.0
 
+    # https://stackoverflow.com/a/73046084
+    Twisted==21.7.0
```

Comment:
for min environment we probably want an earlier twisted, right?

Reply:
Addressed it on a92f771. However, it fails the tests for the SaveFixtureCommand with:

```
2023-02-10 15:45:32 [scrapy.crawler] INFO: Overridden settings:
{'BOT_NAME': 'foo',
 'NEWSPIDER_MODULE': 'foo.spiders',
 'ROBOTSTXT_OBEY': True,
 'SPIDER_MODULES': ['foo.spiders']}
Traceback (most recent call last):
  File "/home/kevinoxy/.pyenv/versions/3.7.2/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/kevinoxy/.pyenv/versions/3.7.2/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/kevinoxy/dev/z/scrapy-poet/.tox/min/lib/python3.7/site-packages/scrapy/cmdline.py", line 168, in <module>
    execute()
  File "/home/kevinoxy/dev/z/scrapy-poet/.tox/min/lib/python3.7/site-packages/scrapy/cmdline.py", line 145, in execute
    _run_print_help(parser, _run_command, cmd, args, opts)
  File "/home/kevinoxy/dev/z/scrapy-poet/.tox/min/lib/python3.7/site-packages/scrapy/cmdline.py", line 100, in _run_print_help
    func(*a, **kw)
  File "/home/kevinoxy/dev/z/scrapy-poet/.tox/min/lib/python3.7/site-packages/scrapy/cmdline.py", line 153, in _run_command
    cmd.run(args, opts)
  File "/home/kevinoxy/dev/z/scrapy-poet/.tox/min/lib/python3.7/site-packages/scrapy_poet/commands.py", line 95, in run
    self.crawler_process.crawl(spider_cls, url=url)
  File "/home/kevinoxy/dev/z/scrapy-poet/.tox/min/lib/python3.7/site-packages/scrapy/crawler.py", line 205, in crawl
    crawler = self.create_crawler(crawler_or_spidercls)
  File "/home/kevinoxy/dev/z/scrapy-poet/.tox/min/lib/python3.7/site-packages/scrapy/crawler.py", line 238, in create_crawler
    return self._create_crawler(crawler_or_spidercls)
  File "/home/kevinoxy/dev/z/scrapy-poet/.tox/min/lib/python3.7/site-packages/scrapy/crawler.py", line 313, in _create_crawler
    return Crawler(spidercls, self.settings, init_reactor=True)
  File "/home/kevinoxy/dev/z/scrapy-poet/.tox/min/lib/python3.7/site-packages/scrapy/crawler.py", line 82, in __init__
    default.install()
  File "/home/kevinoxy/dev/z/scrapy-poet/.tox/min/lib/python3.7/site-packages/twisted/internet/epollreactor.py", line 246, in install
    installReactor(p)
  File "/home/kevinoxy/dev/z/scrapy-poet/.tox/min/lib/python3.7/site-packages/twisted/internet/main.py", line 32, in installReactor
    raise error.ReactorAlreadyInstalledError("reactor already installed")
twisted.internet.error.ReactorAlreadyInstalledError: reactor already installed
```

I think it's due to **Twisted 18.9.0**'s default reactor value.

It could be fixed by setting `self.settings["TWISTED_REACTOR"] = "twisted.internet.epollreactor.EPollReactor"` in SaveFixtureCommand. bf1a40b

I'm not quite knowledgeable about the behavioral specifics of TWISTED\_REACTOR in Scrapy, let me know if there's a better alternative.

Reply:
If I'm reading https://docs.scrapy.org/en/latest/topics/asyncio.html correctly, it means that savefixture won't work with asyncio page objects if we install EPollReactor :( @wRAR is it the case? What would you recommend?

Reply:
> Was the only change the Twisted version downgrade

Yes, this one in particular: a92f771

Reply:
Twisted < 21.2.0 imports (and thus installs) the reactor in `twisted.web.client`, which we import in several places in Scrapy. I don't know yet why this is not a problem for e.g. the `asyncio-pinned` Scrapy tests, maybe some init order is different.

Reply:
Found it: scrapy-poet imports it earlier than Scrapy does, via `scrapy_poet/__init__.py` -> `scrapy_poet/spidermiddlewares.py` -> `scrapy.downloadermiddlewares.retry`.

Moving `from scrapy.downloadermiddlewares.retry import get_retry_request` inside `scrapy_poet.spidermiddlewares.RetryMiddleware.process_spider_exception()` seems to fix this (we could add a comment saying this is needed on Twisted < 21.2.0 only).

Reply:
Awesome work @wRAR ! Let me try to test this out

Reply:
@wRAR Should we make changes in Scrapy to avoid issues like these?

Reply:
@Gallaecio it may be useful to move `twisted.web.client` imports inside methods, if that's actually fixes this problem (so if there are no other things that break this and cannot be fixed).